src/JSObjectProxy.cc

```diff
@@ -0,0 +1,276 @@
+  return;
+}
+
+PyObject *JSObjectProxyMethodDefinitions::JSObjectProxy_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+  JSObjectProxy *self;
+  self = (JSObjectProxy *)type->tp_alloc(type, 0);
+  if (!self)
+  {
+    Py_DECREF(self);
+    return NULL;
+  }
+  self->jsObject.set(JS_NewObject(GLOBAL_CX, NULL));
+  return (PyObject *)self;
+}
+
+int JSObjectProxyMethodDefinitions::JSObjectProxy_init(JSObjectProxy *self, PyObject *args, PyObject *kwds)
+{
+
+  PyObject *dict = NULL;
+  if (PyTuple_Size(args) == 0 || PyTuple_GetItem(args, 0) == Py_None) {
+    // make fresh JSObject for proxy
+    self->jsObject.set(JS_NewObject(GLOBAL_CX, NULL));
+    return 0;
+  }
+
+  if (!PyArg_ParseTuple(args, "O!", &PyDict_Type, &dict))
+  {
+    return -1;
+  }
+
+
+  // make fresh JSObject for proxy
+  self->jsObject.set(JS_NewObject(GLOBAL_CX, NULL));
+  std::unordered_map<PyObject *, JS::RootedValue *> subValsMap;
+  JSObjectProxy_init_helper(self->jsObject, dict, subValsMap);
+  return 0;
+}
+
+void JSObjectProxyMethodDefinitions::JSObjectProxy_init_helper(JS::HandleObject jsObject, PyObject *dict, std::unordered_map<PyObject *, JS::RootedValue *> &subValsMap)
+{
+  PyObject *key, *value;
+  Py_ssize_t pos = 0;
+  while (PyDict_Next(dict, &pos, &key, &value))
+  {
+    bool skip = false;
+    if (!PyUnicode_Check(key))
+    { // only accept string keys
+      continue;
+    }
+
+    for (auto it: subValsMap)
+    {
+      if (it.first == value)
+      { // if we've already seen this value before, just pass the same JS::Value
+        JSObjectProxy_set_helper(jsObject, key, *(it.second));
+        skip = true;
+        break;
+      }
+    }
+
+    if (skip)
+    {
+      continue;
+    }
+
+    JS::RootedValue *jsVal = new JS::RootedValue(GLOBAL_CX, jsTypeFactory(GLOBAL_CX, value));
+    subValsMap.insert({{value, jsVal}});
+    JSObjectProxy_set_helper(jsObject, key, *jsVal);
+  }
+}
+
+Py_ssize_t JSObjectProxyMethodDefinitions::JSObjectProxy_length(JSObjectProxy *self)
+{
+  JS::RootedIdVector props(GLOBAL_CX);
+  if (!js::GetPropertyKeys(GLOBAL_CX, self->jsObject, JSITER_OWNONLY | JSITER_HIDDEN, &props))
+  {
+    // @TODO (Caleb Aikens) raise exception here
+    return -1;
+  }
+
+  return props.length();
+}
+
+PyObject *JSObjectProxyMethodDefinitions::JSObjectProxy_get(JSObjectProxy *self, PyObject *key)
+{
+  if (!PyUnicode_Check(key))
+  {
+    // @TODO (Caleb Aikens) raise exception here
+    return NULL;
+  }
+
+  JS::RootedValue *value = new JS::RootedValue(GLOBAL_CX);
+  switch (PyUnicode_KIND(key))
+  {
+  case PyUnicode_1BYTE_KIND:
+    JS_GetProperty(GLOBAL_CX, self->jsObject, (char *)PyUnicode_1BYTE_DATA(key), value);
+    break;
+  case PyUnicode_2BYTE_KIND:
+    JS_GetUCProperty(GLOBAL_CX, self->jsObject, (char16_t *)PyUnicode_2BYTE_DATA(key), PyUnicode_GET_LENGTH(key), value);
+    break;
+  case PyUnicode_4BYTE_KIND:
+    // @TODO (Caleb Aikens) convert UCS4 to UTF16 and call JS_GetUCProperty
+    break;
+  }
```

Comment:
easier to reuse the code

```suggestion
  
  JS::RootedString idString(GLOBAL_CX);
  switch (PyUnicode_KIND(key))
  {
  case PyUnicode_1BYTE_KIND:
    idString.set(JS_NewStringCopyZ(GLOBAL_CX, (char *)PyUnicode_1BYTE_DATA(key)));
    break;
  case PyUnicode_2BYTE_KIND:
    idString.set(JS_NewUCStringCopyN(GLOBAL_CX, (char16_t *)PyUnicode_2BYTE_DATA(key), PyUnicode_GET_LENGTH(key)));
    break;
  case PyUnicode_4BYTE_KIND:
    // @TODO (Caleb Aikens) convert UCS4 to UTF16 and call JS_GetUCProperty
    break;
  }
  JS::RootedId id(GLOBAL_CX);
  JS_StringToId(GLOBAL_CX, idString, &id);

  JS_GetPropertyById(GLOBAL_CX, self->jsObject, id, value);
```